In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt 
import seaborn as sns

In [2]:
# read the data 
df = pd.read_csv(r"C:\Users\MALWADE TANYA\Downloads\titles.csv")

In [3]:
df.head()

id                                title   type  \
0  ts300399  Five Came Back: The Reference Films   SHOW   
1   tm84618                          Taxi Driver  MOVIE   
2  tm127384      Monty Python and the Holy Grail  MOVIE   
3   tm70993                        Life of Brian  MOVIE   
4  tm190788                         The Exorcist  MOVIE   

                                         description  release_year  \
0  This collection includes 12 World War II-era p...          1945   
1  A mentally unstable Vietnam War veteran works ...          1976   
2  King Arthur, accompanied by his squire, recrui...          1975   
3  Brian Cohen is an average young Jewish man, bu...          1979   
4  12-year-old Regan MacNeil begins to adapt an e...          1973   

  age_certification  runtime                 genres production_countries  \
0             TV-MA       48      ['documentation']               ['US']   
1                 R      113     ['crime', 'drama']               ['US']   
2                PG       91  ['comedy', 'fantasy']               ['GB']   
3                 R       94             ['comedy']               ['GB']   
4                 R      133             ['horror']               ['US']   

   seasons    imdb_id  imdb_score  imdb_votes  tmdb_popularity  tmdb_score  
0      1.0        NaN         NaN         NaN            0.600         NaN  
1      NaN  tt0075314         8.3    795222.0           27.612         8.2  
2      NaN  tt0071853         8.2    530877.0           18.216         7.8  
3      NaN  tt0079470         8.0    392419.0           17.505         7.8  
4      NaN  tt0070047         8.1    391942.0           95.337         7.7

In [4]:
df.shape

(5806, 15)

In [5]:
df.dtypes

id                       object
title                    object
type                     object
description              object
release_year              int64
age_certification        object
runtime                   int64
genres                   object
production_countries     object
seasons                 float64
imdb_id                  object
imdb_score              float64
imdb_votes              float64
tmdb_popularity         float64
tmdb_score              float64
dtype: object

In [6]:
df.describe()

release_year      runtime      seasons   imdb_score    imdb_votes  \
count   5806.000000  5806.000000  2047.000000  5283.000000  5.267000e+03   
mean    2016.013434    77.643989     2.165608     6.533447  2.340719e+04   
std        7.324883    39.474160     2.636207     1.160932  8.713432e+04   
min     1945.000000     0.000000     1.000000     1.500000  5.000000e+00   
25%     2015.000000    44.000000     1.000000     5.800000  5.210000e+02   
50%     2018.000000    84.000000     1.000000     6.600000  2.279000e+03   
75%     2020.000000   105.000000     2.000000     7.400000  1.014400e+04   
max     2022.000000   251.000000    42.000000     9.600000  2.268288e+06   

       tmdb_popularity   tmdb_score  
count      5712.000000  5488.000000  
mean         22.525660     6.818039  
std          68.849177     1.171560  
min           0.009442     0.500000  
25%           3.155250     6.100000  
50%           7.478000     6.900000  
75%          17.775750     7.500000  
max        1823.374000    10.000000

In [7]:
df['type'].value_counts()

type
MOVIE    3759
SHOW     2047
Name: count, dtype: int64

In [8]:
df['release_year'].describe()

count    5806.000000
mean     2016.013434
std         7.324883
min      1945.000000
25%      2015.000000
50%      2018.000000
75%      2020.000000
max      2022.000000
Name: release_year, dtype: float64

In [9]:
df.isna().sum()

id                         0
title                      1
type                       0
description               18
release_year               0
age_certification       2610
runtime                    0
genres                     0
production_countries       0
seasons                 3759
imdb_id                  444
imdb_score               523
imdb_votes               539
tmdb_popularity           94
tmdb_score               318
dtype: int64

***title*** (1 missing )
- one title missing here 
- we can remove these row 

In [10]:
# 3. Drop row where title is missing
df = df.dropna(subset=['title'])

***description*** (18 missing)
- 18 rows missing the description 
- not imp in EDA but 
- u can fill it with 'No Description'

In [11]:
# 1. Fill age_certification
df['description'] = df['description'].fillna('No Description')

***seasons*** (3759 missing)
- movies not have seasons 
- movies = missing sessions
- we can fill it using 0

In [12]:
# 2. Fill seasons (movies = 0)
df['seasons'] = df['seasons'].fillna(0)

***age_certification*** (2610 missing)
- these are the missing age groups 
- if movies or serirs are so old or international containt 
- we can fill it using 'Not Rated' using fill na

In [13]:
# 1. Fill age_certification
df['age_certification'] = df['age_certification'].fillna('Not Rated')

***imdb_id*** (444 missing) but 
- we don't use id column in analysis 
- it is only use for identity 
- so we can drop this column 

In [14]:
df.drop(columns=['imdb_id'], inplace=True)

***imdb_score*** (522 Missing)
- means IMDb rating (0–10)
- not used in EDA but can be imp for ML
- we can fill it using median of imdb_score

In [15]:
df['imdb_score'].fillna(df['imdb_score'].median(), inplace=True)

C:\Users\MALWADE TANYA\AppData\Local\Temp\ipykernel_20072\4214535340.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['imdb_score'].fillna(df['imdb_score'].median(), inplace=True)


***imdb_votes*** (538 missing)
- use to understand the popularity 
- not used in EDA but important for ML
- fill with median 

In [16]:
df['imdb_votes'].fillna(df['imdb_votes'].median(), inplace=True)

C:\Users\MALWADE TANYA\AppData\Local\Temp\ipykernel_20072\3786480325.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['imdb_votes'].fillna(df['imdb_votes'].median(), inplace=True)


- both are numerical colums we can fill it with median

In [17]:
df['tmdb_popularity'].fillna(df['tmdb_popularity'].median(), inplace=True)
df['tmdb_score'].fillna(df['tmdb_score'].median(), inplace=True)

C:\Users\MALWADE TANYA\AppData\Local\Temp\ipykernel_20072\2600862408.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['tmdb_popularity'].fillna(df['tmdb_popularity'].median(), inplace=True)
C:\Users\MALWADE TANYA\AppData\Local\Temp\ipykernel_20072\2600862408.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting

In [18]:
df.isna().sum()

id                      0
title                   0
type                    0
description             0
release_year            0
age_certification       0
runtime                 0
genres                  0
production_countries    0
seasons                 0
imdb_score              0
imdb_votes              0
tmdb_popularity         0
tmdb_score              0
dtype: int64

#### Learnings
- Real-world datasets are often messy and require careful data cleaning.
- Missing values are not always errors; sometimes they provide useful information (e.g., missing seasons indicating a movie).
- Data cleaning involves logical decision-making, not just deleting rows or columns.
- Median is more reliable than mean for filling missing values in skewed numerical data.
- Understanding the meaning of each column is the foundation of effective exploratory data analysis.